### PR TITLE
TFP: lr=8e-5 upper bracket probe (new champion was 7.5e-5)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Map the upper bound of the TFP LR optimum. Your lr=7.5e-5 result (#3352) crushed the lr=7e-5 baseline by -14.8% val and -7.4% test. The LR curve may continue upward — lr=8e-5 is the next point to check. If lr=8e-5 improves further, the optimum is above 7.5e-5. If it's worse, 7.5e-5 is confirmed as the peak.

Also run lr=7.25e-5 as a secondary trial to narrow the bracket between 7e-5 and 7.5e-5.

## Current Baseline (TandemFoil Paper) — JUST UPDATED

- **val_primary/field_mse:** 0.001857 at epoch 490 (your lr=7.5e-5 run, PR #3352)
- **test_primary/field_mse:** 0.001789

## Instructions to Student

Run both trials in parallel:

**Trial 1 (lr=8e-5):**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 8e-5 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --eval-interval 10 \
  --save-checkpoint \
  --seed 0 \
  --wandb_group tfp-lr-bracket \
  --wandb_name "haku/tfp-lr8e5"
```

**Trial 2 (lr=7.25e-5):**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 7.25e-5 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --eval-interval 10 \
  --save-checkpoint \
  --seed 0 \
  --wandb_group tfp-lr-bracket \
  --wandb_name "haku/tfp-lr725e5"
```

**What to report for each trial:**
1. Best `val_primary/field_mse` and `test_primary/field_mse` and epoch
2. W&B run IDs
3. Val/surface_mse and val/volume_mse breakdown

**Target to beat:** val < 0.001857 (your own new baseline from #3352)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### Trial 1: lr=8e-5

| Metric | Baseline (lr=7.5e-5) | This run | Delta |
|--------|---------------------|----------|-------|
| val_primary/field_mse | 0.001857 | **0.002002** | +7.8% ❌ |
| test_primary/field_mse | 0.001789 | **0.001921** | +7.4% ❌ |
| val/surface_mse | — | 0.001245 | — |
| val/volume_mse | — | 0.002014 | — |

- **Best epoch:** 550 (of 577 trained)
- **W&B run:** `o7w5rex5`
- **Training:** Stable throughout. grad_norm_mean ~5-13, loss steadily decreased. No instability.
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 8e-5 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition --epochs 999 \
  --ema-decay 0.999 --eval-interval 10 --save-checkpoint --seed 0 \
  --agent haku --wandb-group tfp-lr-bracket --wandb-name "haku/tfp-lr8e5"
```

### Trial 2: lr=7.25e-5

| Metric | Baseline (lr=7.5e-5) | This run | Delta |
|--------|---------------------|----------|-------|
| val_primary/field_mse | 0.001857 | **0.003901** | +110% ❌ |
| test_primary/field_mse | 0.001789 | **0.004026** | +125% ❌ |
| val/surface_mse | — | 0.001825 | — |
| val/volume_mse | — | 0.003934 | — |

- **Best epoch:** 180 (of 579 trained)
- **W&B run:** `3d2lqnnc`
- **Training:** **Diverged catastrophically.** Val improved to 0.003901 by epoch 180, then started oscillating at epoch 190. By epoch 400 val exploded to 1.72. Gradients went fully NaN at epoch 474 (all 360 micro-batches non-finite). Loss settled at ~2.88 for the remaining epochs. Even before divergence, val was 2x worse than baseline.
- **Command:** Same as Trial 1 with `--lr 7.25e-5` and `--wandb-name "haku/tfp-lr725e5"`

## What happened

**Trial 1 (lr=8e-5):** Confirms the upper boundary of the LR optimum. lr=8e-5 trained stably but is ~7.5% worse than lr=7.5e-5 on both val and test. The LR curve turns downward above 7.5e-5, confirming **7.5e-5 is at or near the peak**.

**Trial 2 (lr=7.25e-5):** Surprising result — this LR is *between* the old baseline (7e-5) and the champion (7.5e-5), yet it diverged badly. Even before divergence, best val (0.003901) was far worse than both neighbors. This suggests the LR landscape may not be smooth, or there's a specific interaction between this LR value and the cosine-T_max=10 schedule that creates instability. The cosine restarts may hit a resonance at this particular learning rate.

## Suggested follow-ups
1. **Seed robustness for lr=7.25e-5** — re-run with seed=42 to check if the divergence is seed-specific or structural
2. **Fine-grained upper bracket** — try lr=7.75e-5 to narrow the gap between the champion (7.5e-5) and the confirmed-worse 8e-5
3. **Investigate cosine schedule interaction** — the divergence pattern (stable for 180 epochs, then collapse) suggests the cosine restarts may accumulate instability at certain LR values. A longer T_max (e.g., 15 or 20) might help at lr=7.25e-5

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3377 because ADVISOR: Closing — both LR bracket probes confirmed 7.5e-5 is the TFP optimum. lr=8e-5: val=0.002002, test=0.001921 (+7.5% worse). lr=7.25e-5: DIVERGED catastrophically (val→1.72 by ep400, NaN at ep474). The LR landscape is surprisingly non-smooth — 7.25e-5 sits in an unstable zone between 7e-5 and 7.5e-5. The 4L direction (yuji test=0.001712) is now the most promising TFP frontier..
